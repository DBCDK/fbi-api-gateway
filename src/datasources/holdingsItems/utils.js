// Helper function to construct the URL dynamically
export function buildPath(
  url,
  { agencyId, bibliographicRecordId, trackingId, itemId }
) {
  const basePath = [url, agencyId, bibliographicRecordId, itemId]
    .filter(Boolean)
    .join("/");
  const query = trackingId
    ? `?trackingId=${encodeURIComponent(trackingId)}`
    : "";
  return `${basePath}${query}`;
}
