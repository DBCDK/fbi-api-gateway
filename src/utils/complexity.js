import { log } from "dbc-node-logger";
import { parse, GraphQLError } from "graphql";

import queryComplexity, {
  getComplexity,
  simpleEstimator,
  directiveEstimator,
} from "graphql-query-complexity";

import config from "../config";

/**
 * Function to build the content for the estimator functions
 *
 * @param {*} params.query
 * @param {*} params.variables
 * @param {object} params.schema
 * @returns {object}
 */
export function buildQueryComplexity({
  schema = null,
  query,
  variables,
  maxQueryComplexity = null,
}) {
  // Set the complexity limit (custom or default/config)
  const COMPLEXITY_LIMIT = maxQueryComplexity || config.query.maxComplexity;

  return {
    estimators: [
      directiveEstimator(),
      simpleEstimator({
        defaultComplexity: 1,
      }),
    ],
    schema,
    maximumComplexity: COMPLEXITY_LIMIT,
    query: parse(query),
    variables,
    createError: (max, actual) => {
      return new GraphQLError(
        `Query is too complex: ${actual}. Maximum allowed complexity: ${max}`
      );
    },
    onComplete: (complexity) => {
      if (complexity > COMPLEXITY_LIMIT) {
        log.error("Query exceeded complexity limit", {
          complexity,
          query,
          clientConfiguredLimit: !!maxQueryComplexity,
          maxComplexity: COMPLEXITY_LIMIT,
        });
      }
    },
  };
}

/**
 * Future complexity estimator (returns validation function)
 *
 * @param {*} params.query
 * @param {*} params.variables
 * @param {object} params.schema
 * @returns {funciton}
 */

export function validateQueryComplexity(params) {
  try {
    return queryComplexity(buildQueryComplexity(params));
  } catch (e) {
    // Log error in case complexity cannot be calculated (invalid query, misconfiguration, etc.)
  }
}

/**
 * Future complexity estimator (returns complexity value)
 *
 * @param {*} params.query
 * @param {*} params.variables
 * @param {object} params.schema
 * @returns {int}
 */

export function getQueryComplexity(params) {
  try {
    return getComplexity(buildQueryComplexity(params));
  } catch (e) {
    // Log error in case complexity cannot be calculated (invalid query, misconfiguration, etc.)
  }
}

/**
 *
 * @param {int} complexity
 * @returns {string}
 */
export function getQueryComplexityClass(complexity) {
  if (complexity < 250) {
    return "simple";
  }
  if (complexity < 2500) {
    return "complex";
  }
  if (complexity < 25000) {
    return "critical";
  }

  return "rejected";
}
