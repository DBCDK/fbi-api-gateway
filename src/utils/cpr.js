/**
 * Check for validity of a cpr
 *
 * @param cpr
 * @returns {boolean|Date|boolean}
 */
export function isValidCpr(cpr) {
  return isNumeric(cpr) && isValidDate(cpr) && cpr.length === 10;
}

/**
 *
 * @param n
 * @returns {boolean}
 */
function isNumeric(n) {
  return !isNaN(parseFloat(n)) && isFinite(n);
}

/**
 * Check if first 6 digits contains a valid date
 * @param ddmmyy
 * @returns {Date|boolean}
 */
function isValidDate(ddmmyy) {
  const mm = ddmmyy.substr(2, 2) || "00";
  const d = new Date(ddmmyy.substr(4, 2), mm - 1, ddmmyy.substr(0, 2));
  return d && d.getMonth() + 1 === parseInt(mm, 10);
}
