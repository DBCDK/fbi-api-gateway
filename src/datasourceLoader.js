import DataLoader from "dataloader";

import { log } from "dbc-node-logger";
import { withRedis, clearRedis } from "./datasources/redis.datasource";
import { createFetchWithConcurrencyLimit } from "./utils/fetchWithLimit";
import { getFilesRecursive } from "./utils/utils";
import config from "./config";

/*
  This is how we time parallel requests to a single datasource.
  
  Request A
  ------------
          Request B
          ------------------
                                      Request C
                                      -----------
  Total AB
  --------------------------
  
  The totaltime for the requests should NOT be:
  Request A + Request B + Request C
  
  But instead:
  Total AB + Request C
*/

export const trackMe = () => {
  return {
    start: function (uuid) {
      this.uuid = uuid;
      return this;
    },
    createTracker: function (name) {
      // If this datasource has not been tracked yet, initialize its tracking object.
      if (!this.trackObject[name]) {
        this.trackObject[name] = {
          overlappingRequestCount: 0, // Number of ongoing requests for this datasource.
          count: 0, // Total number of elements retrieved from datasource
          started: 0, // Timestamp when the first concurrent request starts.
          time: 0, // Total time taken for all requests so far.
          sequentialTime: 0, // If requests were to be run sequentially, this is the total time
          batches: 0, // Dataloader batches multiple requests, into a single request when possible
        };
      }
    },

    // Function to indicate the beginning of a datasource call.
    begin: function (name, count) {
      this.createTracker(name);

      const beginTime = performance.now();

      const trackObject = this.trackObject[name];

      // Increment the count of ongoing requests.
      trackObject.overlappingRequestCount++;

      // Increment total number of elements retrieved from datasource
      trackObject.count += count;
      trackObject.batches++;

      // Function to indicate the end of an datasource call.
      return function end() {
        const endTime = performance.now();
        const duration = Math.round(endTime - beginTime);
        // Decrement the count of ongoing requests.
        trackObject.overlappingRequestCount--;

        // If all ongoing requests are complete, calculate the time taken.
        if (trackObject.overlappingRequestCount === 0) {
          trackObject.time += duration;
        }

        trackObject.sequentialTime += duration;
      };
    },

    cacheMiss: function (name, count) {
      this.createTracker(name);

      this.trackObject[name].cacheMiss =
        (this.trackObject[name].cacheMiss || 0) + count;
    },
    uuid: null,
    trackObject: {},
    getMetrics: function () {
      const res = {};

      Object.entries(this.trackObject).forEach(([key, val]) => {
        res[key] = {
          count: val.count,
          time: val.time,
          cacheMiss: val.cacheMiss,
          avg_ms: Math.round(val.sequentialTime / val.count),
        };
      });

      return res;
    },
  };
};

// Find all datasources in src/datasources
export const datasources = getFilesRecursive(`${__dirname}/datasources`)
  .map((file) => {
    if (!file.path.endsWith(".datasource.js")) {
      return;
    }
    const { load, options, batchLoader, status, testLoad } = require(file.path);
    if (!load && !batchLoader) {
      return;
    }

    // Extract datasource name from filename
    const name = file.file.replace(".datasource.js", "");

    return {
      testLoad,
      load,
      batchLoader,
      name,
      options,
      statusChecker: status ? () => status(load) : null,
    };
  })
  .filter((func) => !!func);

const nameToDatasource = {};
datasources.forEach((ds) => (nameToDatasource[ds.name] = ds));

log.debug(
  `found ${datasources.length} datasources, ${datasources
    .map((datasource) => datasource.name)
    .join(", ")}`
);

/**
 * This will initialize a single dataloader.
 * It makes sure a context (with a concurrency limited fetcher), is injected
 * into the batchloader function.
 *
 * It sets up Redis for the dataloader
 *
 * It collects performance metrics for the datasource.
 */
function setupDataloader(
  { name, load, options, batchLoader, testLoad },
  context
) {
  const testModeActivated = context?.testUser && testLoad;
  const loadImpl = testModeActivated ? testLoad : load;

  let batchLoaderWithContext = batchLoader
    ? (keys) => batchLoader(keys, context)
    : (keys) => Promise.all(keys.map((key) => loadImpl(key, context)));

  if (!testModeActivated && options?.redis?.prefix && options?.redis?.ttl) {
    batchLoaderWithContext = withRedis(batchLoaderWithContext, {
      ...options.redis,
      ...context,
      datasourceName: name,
    });
  }

  async function batchLoaderWithTiming(keys) {
    const end = context?.track?.begin?.(name, keys.length);

    try {
      return await batchLoaderWithContext(keys);
    } catch (err) {
      // Wraps the load call in a try catch, in order to capture the
      // stack trace before dataloader is called. Otherwise we would
      // be limited to see the stack trace from the data sources, but
      // not the GraphQL resolvers.
      // https://brunoscheufler.com/blog/2021-07-25-improving-dataloader-stack-traces
      if (err instanceof Error && err.stack) {
        const currentStack = { stack: "" };
        Error.captureStackTrace(currentStack, batchLoaderWithTiming);
        err.stack += currentStack.stack.replace("Error", "\nBefore DataLoader");
      }
      throw err;
    } finally {
      end?.();
    }
  }

  const loader = new DataLoader(batchLoaderWithTiming, {
    // If key is an object, we stringify
    // to make it useful as a cache key
    cacheKeyFn: (key) => (typeof key === "object" ? JSON.stringify(key) : key),
    maxBatchSize: 100,
  });

  if (options?.redis?.prefix) {
    loader.clearRedis = (key) => clearRedis(options?.redis?.prefix, key);
  }

  return {
    loader,
  };
}

/**
 * Will instantiate dataloaders from datasources.
 * This should be done for every incoming GraphQL request
 */
export default function createDataLoaders(uuid, testUser, accessToken) {
  const result = {};
  const track = trackMe().start(uuid);

  result.trackingObject = track;
  const fetchWithConcurrencyLimit = createFetchWithConcurrencyLimit(
    config.fetchConcurrencyLimit
  );
  // Gets a loader by name.
  // A loader will be initialized first time it is called
  function getLoader(name) {
    if (!result[name]) {
      result[name] = setupDataloader(nameToDatasource[name], {
        getLoader,
        track,
        fetch: (url, options) => fetchWithConcurrencyLimit(url, options, name),
        trackingId: uuid,
        testUser,
        accessToken,
      })?.loader;
    }
    return result[name];
  }
  result.getLoader = getLoader;
  return result;
}

/**
 * For testing
 */
export function createMockedDataLoaders() {
  const mockedDatasources = {};
  getFilesRecursive("./src/datasources/mocked")
    .filter(
      (file) =>
        file.path.endsWith("datasource.mocked.js") && require(file.path).load
    )
    .map((file) => ({
      ...file,
      name: file.file.replace(".datasource.mocked.js", ""),
      load: require(file.path).load,
      clearRedis: () => {},
    }))
    .forEach((loader) => {
      mockedDatasources[loader.name] = loader;
    });

  mockedDatasources.getLoader = (name) => mockedDatasources[name];

  return mockedDatasources;
}

export function createTestUserDataLoaders() {
  const mockedDataLoaders = createMockedDataLoaders();
  getFilesRecursive("./src/datasources")
    .filter(
      (file) =>
        file.path.endsWith("datasource.js") && require(file.path).testLoad
    )
    .map((file) => ({
      ...file,
      name: file.file.replace(".datasource.js", ""),
      load: require(file.path).testLoad,
      clearRedis: () => {},
    }))
    .forEach((loader) => {
      mockedDataLoaders[loader.name] = loader;
    });

  return mockedDataLoaders;
}
