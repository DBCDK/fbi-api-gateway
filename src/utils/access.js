/**
 * @file access util functions
 */

import { filterDuplicateAgencies, resolveManifestation } from "./utils";

/**
 *
 * Check whether any of the users libraries has access to Infomedia.
 * The returned agencyId will be billed for infomedia access.
 *
 * First we check if the users login library has access to info media, if not,
 * we check if any other og the users associated accounts has access to infomedia,
 * if so, we return the first one found.
 *
 * @param {object} context
 *
 * @returns {null | string} AgencyId
 */
async function infomedia(context) {
  const user = context?.user;

  // get rights from idp
  const idpRights = await context.datasources.getLoader("idp").load("");

  // check if users loggedInAgency has infomedia access
  const loggedInAgencyId = user?.loggedInAgencyId;

  if (loggedInAgencyId && idpRights[loggedInAgencyId]) {
    return loggedInAgencyId;
  }

  /*
   * Alternativly check all users accounts
   */

  // filtered accounts
  const userInfoAccounts = filterDuplicateAgencies(user.agencies);

  const hasAccess = userInfoAccounts?.filter(
    ({ agencyId }) => idpRights[agencyId]
  );

  // check for infomedia access - if any of users agencies subscribes
  if (hasAccess.length > 0) {
    return hasAccess[0]?.agencyId;
  }

  return null;
}

/**
 *
 * @param {object} context
 *
 * @returns {null | string} AgencyId
 */
export async function getInfomediaAgencyId(context) {
  return await infomedia(context);
}

/**
 *
 * @param {object} context
 *
 * @returns {boolean}
 */
export async function hasInfomediaAccess(context) {
  return !!(await infomedia(context));
}

/**
 * Get type of access for manifestation with given pid.
 *  url access
 *  ill access
 *  infomedia access
 *  digital article service access
 *
 * @param pid
 * @param context
 * @returns {Promise<*>}
 */
export async function resolveAccess(manifestation, context) {
  // We parse the access structure from JED, and convert it
  // to the union type structure.
  // At some point we may choose to follow the structure of JED closely,
  // but it has to be coordinated with stakeholders

  let parent =
    typeof manifestation === "object"
      ? manifestation
      : await resolveManifestation({ pid: manifestation }, context);

  const res = [];

  // use linkchecker to check status of a single accessUrl
  // linkchecker does NOT handle proxy urls, so we skip ebookcentral and ebsco urls
  const linkStatus = async (url) => {
    if (url?.includes("ebookcentral") || url?.includes("ebscohost")) {
      return "OK";
    }

    const linkcheck = await context.datasources
      .getLoader("linkcheck")
      .load({ urls: [url] });
    const link = linkcheck.find((check) => check.url === url);

    return link?.status || "OK";
  };

  parent?.access?.accessUrls?.forEach((entry) => {
    const { proxyUrl, loginRequired } = getProxyUrl(
      entry.url || "",
      context?.user
    );

    res.push({
      __typename: "AccessUrl",
      origin: parseOnlineUrlToOrigin(entry.url),
      url: proxyUrl,
      loginRequired,
      note: entry.note,
      type: entry.type,
      status: linkStatus(proxyUrl) || "OK",
    });
  });

  if (parent?.access?.dbcWebArchive) {
    const archives = await context.datasources
      .getLoader("moreinfoWebarchive")
      .load(parent.pid);

    archives.forEach((archive) => {
      if (archive.url) {
        res.push({
          __typename: "AccessUrl",
          url: archive.url,
          origin: parseOnlineUrlToOrigin(archive.url),
          loginRequired: false,
          type: "RESOURCE",
          status: linkStatus(archive.url) || "OK",
        });
      }
    });
  }

  parent?.access?.ereol?.forEach((entry) => {
    res.push({
      __typename: "Ereol",
      origin: entry.origin,
      url: entry.url,
      canAlwaysBeLoaned: entry.canAlwaysBeLoaned,
      note: entry.note,
    });
  });

  if (parent?.access?.infomediaService?.id) {
    if (
      !["politiken", "jyllands-posten"].some((publication) =>
        parent?.hostPublication?.title?.toLowerCase()?.includes(publication)
      )
    ) {
      res.push({
        __typename: "InfomediaService",
        id: parent.access.infomediaService.id,
      });
    }

    // If this is a omitted online access for politiken or jyllands-posten.
    else {
      const hasPhysical = parent?.accessTypes?.some(
        (t) => (t?.code ?? "").toUpperCase() === "PHYSICAL"
      );

      if (hasPhysical) {
        parent = {
          ...parent,
          access: {
            ...(parent.access ?? {}),
            interLibraryLoanIsPossible: true,
          },
        };
      }
    }
  }

  // Get the issn for this article or periodica
  const issn =
    parent?.access?.digitalArticleService?.issn ||
    parent?.identifiers?.find?.((id) => id.type === "ISSN")?.value;

  if (issn) {
    const journals = await context.datasources
      .getLoader("statsbiblioteketJournals")
      .load("");
    const articleissn = issn.replace(/[^a-z\d]/gi, "");
    const hasJournal = journals && journals[articleissn];
    if (hasJournal) {
      res.push({
        __typename: "DigitalArticleService",
        issn: articleissn,
      });
    }
  }

  if (parent?.access?.interLibraryLoanIsPossible) {
    //if collectionidentifiers includes 870970-accessnew it may be loaned
    const isNewMaterial =
      parent?.collectionIdentifiers?.includes("870970-accessnew");

    const forLoan = checkInterLibraryLoan(parent, context);
    res.push({
      __typename: "InterLibraryLoan",
      loanIsPossible: forLoan,
      accessNew: isNewMaterial,
    });
  }

  // Return array containing all types of access
  return _sortOnlineAccess(res);
}

/**
 * municipality number is the second|third|fourth digit in agencyId
 * @param agencyId
 * @returns {string | undefined}
 */
function parseForMunicipalityNumber(agencyId) {
  return agencyId?.substring(1, 4);
}

/**
 * This one is for ebook.plus - we need to go via a proxy url (if user is logged in)
 * @param url
 * @param user
 * @returns {*}
 */
export function getProxyUrl(url, user) {
  const municipality =
    user?.municipality ||
    parseForMunicipalityNumber(user?.municipalityAgencyId);

  // check if we should proxy this url - for now it is ebookcentral and ebscohost
  const proxyMe =
    url.indexOf("ebookcentral") !== -1 || url.indexOf("ebscohost") !== -1;
  if (proxyMe) {
    // check if user is logged in
    if (user?.userId) {
      const realUrl = `https://bib${municipality}.bibbaser.dk/login?url=${url}`;
      return { proxyUrl: realUrl, loginRequired: proxyMe };
    }
  }

  return { proxyUrl: url, loginRequired: proxyMe };
}

/**
 * Some sources are not for loan (fagbibliografier) - for now a statis list:
 * This method is only called if interLibraryLoan is true - We know that for a fact
 * 159080-fagbib - Besættelsesbibliografien
 * 159081-fagbib - Bibliografi over Dansk Kunst
 * 159082-fagbib - Dansk Historisk Bibliografi
 * 159083-fagbib - Dansk Musiklitterær Bibliografi
 * 159084-fagbib - Kongelige Teater programartikler
 * 159085-fagbib - Dania Polyglotta
 * 159086-fagbib - Sportline
 * @param parent
 * @param context
 */
function checkInterLibraryLoan(parent, context) {
  // This is a blacklist of bases NOT to be loaned
  const notForLoanList = [
    "159080-fagbib",
    "159081-fagbib",
    "159082-fagbib",
    "159083-fagbib",
    "159084-fagbib",
    "159085-fagbib",
    "159086-fagbib",
  ];
  // we check on objectId - first part is the source
  const source = parent?.objectId?.split(":")[0];
  return !notForLoanList.includes(source);
}

/**
 * Get domain from url
 * @param url
 * @return String
 *  A parsed string eg. "DBC Webarkiv" or name of host eg. "infolink2003.elbo.dk"
 *
 * */
export function parseOnlineUrlToOrigin(url) {
  try {
    const parsedUrl = new URL(url);
    if (parsedUrl["host"] === "moreinfo.addi.dk") {
      return "DBC Webarkiv";
    } else {
      return (parsedUrl["host"] && parsedUrl["host"]) || "";
    }
  } catch (e) {
    return "";
  }
}

/**
 * Handle special cases - for now we want filmstriben/fjernleje BEFORE filmstriben/biblioteket AND
 * we want dansk film institut (dfi) to come last - it is not a 'real' url but a link to a description
 * @param onlineAccess
 * @return {*}
 * @private
 */
function _sortOnlineAccess(onlineAccess) {
  const specialSort = (a, b) => {
    // fjernleje should be on top
    if (b.url && b.url.indexOf("filmstriben.dk/fjernleje") !== -1) {
      return 1;
    } else if (a.url && a.url.indexOf("filmstriben.dk/fjernleje") !== -1) {
      return -1;
      // dfi is not a 'real' online url - sort low
    } else if (b.url && b.url.indexOf("dfi.dk") !== -1) {
      return -1;
    } else if (a.url && a.url.indexOf("dfi.dk") !== -1) {
      return 1;
    }
    return 0;
  };

  return onlineAccess.sort(specialSort);
}
