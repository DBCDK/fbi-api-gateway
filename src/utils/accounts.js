/**
 * @file This file handles util function to select single or multiple CULR accounts
 *
 */

import { isValidCpr } from "./cpr";

/**
 *
 * @param {string} accessToken
 * @param {object} context
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Promise} Object or null
 */
export async function getAccount(accessToken, context, props) {
  return (await getAccounts(accessToken, context, props))?.[0] || null;
}

/**
 *
 * @param {string} accessToken
 * @param {object} context
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Promise} (Array)
 */
export async function getAccounts(accessToken, context, props) {
  // userInfo
  const userinfo = await context.datasources.getLoader("userinfo").load({
    accessToken,
  });

  const user = userinfo?.attributes;

  if (!user?.userId) {
    return [];
  }

  // select dataloader
  let dataloader = isValidCpr(user.userId)
    ? "culrGetAccountsByGlobalId"
    : "culrGetAccountsByLocalId";

  // Retrieve user culr account
  const response = await context.datasources.getLoader(dataloader).load({
    userId: user.userId,
    agencyId: user.loggedInAgencyId,
  });

  return filterAccountsByProps(response.accounts, props);
}

/**
 *
 * @param {Array} accounts
 * @param {object} props - Filter accounts by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Array}
 */
export function filterAccountsByProps(accounts = [], props = {}) {
  if (props.agency) {
    accounts = accounts?.filter(({ agencyId }) => agencyId === props.agency);
  }
  if (props.id) {
    accounts = accounts?.filter(({ userIdValue }) => userIdValue === props.id);
  }
  if (props.type) {
    accounts = accounts?.filter(({ userIdType }) => userIdType === props.type);
  }
  return accounts;
}

/**
 *
 * Function to select a specific agency from a userinfo agencies list
 *
 * @param {Array} agencies
 * @param {object} props - Filter agencies by given props (optional)
 * @param {string} props.id (optional)
 * @param {string} props.agency (optional)
 * @param {string} props.type - (optional)
 *
 * @returns {Array}
 */
export function filterAgenciesByProps(agencies = [], props = {}) {
  if (props.agency) {
    agencies = agencies?.filter(({ agencyId }) => agencyId === props.agency);
  }
  if (props.id) {
    agencies = agencies?.filter(({ userId }) => userId === props.id);
  }
  if (props.type) {
    agencies = agencies?.filter(({ userIdType }) => userIdType === props.type);
  }
  return agencies;
}
