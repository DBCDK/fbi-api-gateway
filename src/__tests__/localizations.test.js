import { createMockedDataLoaders } from "../datasourceLoader";
import { performTestQuery } from "../utils/utils";

test("localizations - get for a number of pids", async () => {
  const result = await performTestQuery({
    query: `
          query  {
          localizations(pids: ["870970-basis:29433909"]) {
            count
            agencies {
              agencyId
              holdingItems {
                localizationPid
                localIdentifier
              }
            }
          }
        }
        `,
    variables: {},
    context: { datasources: createMockedDataLoaders() },
  });
  expect(result).toMatchSnapshot();
});
