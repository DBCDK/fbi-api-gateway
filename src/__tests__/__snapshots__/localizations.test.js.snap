// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`localizations - get for a number of pids 1`] = `
Object {
  "data": Object {
    "localizations": Object {
      "agencies": Array [
        Object {
          "agencyId": "785100",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "775100",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "716500",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "720100",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "715100",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "781300",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "714700",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "754000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "781000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "911116",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "700400",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "800010",
          "holdingItems": Array [
            Object {
              "localIdentifier": "99122450879205763__1",
              "localizationPid": "800010-katalog:99122450879205763__1",
            },
            Object {
              "localIdentifier": "99122160750405763__1",
              "localizationPid": "800010-katalog:99122160750405763__1",
            },
            Object {
              "localIdentifier": "99122450879205763__1",
              "localizationPid": "800010-katalog:99122450879205763__1",
            },
          ],
        },
        Object {
          "agencyId": "715500",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "724000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "732000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "852470",
          "holdingItems": Array [
            Object {
              "localIdentifier": "93142861",
              "localizationPid": "852470-katalog:93142861",
            },
          ],
        },
        Object {
          "agencyId": "721000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "760700",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "725000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "719000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "715700",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "715900",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "725300",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "727000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "751000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "726000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "776600",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "721700",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "716300",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "766100",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "716700",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "716900",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "775600",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "732600",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "762100",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "718300",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "710100",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "717300",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "784600",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "741000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "876040",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "745000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "730600",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "773000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "784000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "732900",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "723000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "774100",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "774600",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "733000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "734000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "763000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "779100",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
        Object {
          "agencyId": "739000",
          "holdingItems": Array [
            Object {
              "localIdentifier": "29433909",
              "localizationPid": "870970-basis:29433909",
            },
          ],
        },
      ],
      "count": 54,
    },
  },
}
`;
